stable_baselines3/common/buffers.py

```diff
@@ -381,3 +381,118 @@ def _get_samples(self, batch_inds: np.ndarray, env: Optional[VecNormalize] = Non
+
+
+class NstepReplayBuffer(ReplayBuffer):
+    """
+    Replay Buffer that computes N-step returns.
+
+    :param buffer_size: (int) Max number of element in the buffer
+    :param observation_space: (spaces.Space) Observation space
+    :param action_space: (spaces.Space) Action space
+    :param device: (Union[th.device, str]) PyTorch device
+        to which the values will be converted
+    :param n_envs: (int) Number of parallel environments
+    :param optimize_memory_usage: (bool) Enable a memory efficient variant
+        of the replay buffer which reduces by almost a factor two the memory used,
+        at a cost of more complexity.
+        See https://github.com/DLR-RM/stable-baselines3/issues/37#issuecomment-637501195
+        and https://github.com/DLR-RM/stable-baselines3/pull/28#issuecomment-637559274
+    :param n_step: (int) The number of transitions to consider when computing n-step returns
+    :param gamma:  (float) The discount factor for future rewards.
+    """
+
+    def __init__(
+        self,
+        buffer_size: int,
+        observation_space: spaces.Space,
+        action_space: spaces.Space,
+        device: Union[th.device, str] = "cpu",
+        n_envs: int = 1,
+        optimize_memory_usage: bool = False,
+        n_step: int = 1,
+        gamma: float = 0.99,
+    ):
+        super().__init__(buffer_size, observation_space, action_space, device, n_envs, optimize_memory_usage)
+        self.n_step = int(n_step)
+        if not 0 < n_step <= buffer_size:
+            raise ValueError("n_step needs to be strictly smaller than buffer_size, and strictly larger than 0")
+        self.gamma = gamma
+
+    def _get_samples(self, batch_inds: np.ndarray, env: Optional[VecNormalize] = None) -> ReplayBufferSamples:
+        # TODO(PartiallyTyped): explain why this assert was here
+        # and check why it can fail (it fails during some tests with DQN)
+        # assert not np.any(batch_inds == self.pos)
+        actions = self.actions[batch_inds, 0, :]
+
+        gamma = self.gamma
+
+        # Broadcasting turns 1dim arange matrix to 2 dimensional matrix that contains all
+        # the indices, % buffersize keeps us in buffer range
+        # indices is a [B x n_step ] matrix
+        indices = (np.arange(self.n_step) + batch_inds.reshape(-1, 1)) % self.buffer_size
+
+        # two dim matrix of not dones. If done is true, then subsequent dones are turned to 0
+        # using accumulate. This ensures that we don't use invalid transitions
+        # not_dones is a [B x n_step] matrix
+        not_dones = np.squeeze(np.multiply.accumulate(1 - self.dones[indices], 1), -1)
+        # vector of the discount factors
+        # [n_step] vector
+        gammas = gamma ** np.arange(self.n_step)
+
+        # two dim matrix of rewards for the indices
+        # using indices we select the current transition, plus the next n_step ones
+        rewards = np.squeeze(self.rewards[indices], -1)
```

Comment:
minor: please always specify explicitly `axis=` so we know what is that parameter